File: src/content/docs/topics/cs/design-patterns/interpreter.py
from __future__ import annotations
from typing import Protocol


Context = dict[str, bool]


class AbstractExpression(Protocol):
    def interpret(self, context: Context) -> bool: ...


class Variable:
    def __init__(self, name: str) -> None:
        self._name = name

    def interpret(self, context: Context) -> bool:
        return context.get(self._name, False)


class AndExpression:
    def __init__(self, left: AbstractExpression, right: AbstractExpression) -> None:
        self._left = left
        self._right = right

    def interpret(self, context: Context) -> bool:
        return self._left.interpret(context) and self._right.interpret(context)


class OrExpression:
    def __init__(self, left: AbstractExpression, right: AbstractExpression) -> None:
        self._left = left
        self._right = right

    def interpret(self, context: Context) -> bool:
        return self._left.interpret(context) or self._right.interpret(context)


class NotExpression:
    def __init__(self, operand: AbstractExpression) -> None:
        self._operand = operand

    def interpret(self, context: Context) -> bool:
        return not self._operand.interpret(context)


# (x AND y) OR (NOT z)
x, y, z = Variable('x'), Variable('y'), Variable('z')
expr = OrExpression(AndExpression(x, y), NotExpression(z))

ctx = {'x': True, 'y': False, 'z': False}
print(expr.interpret(ctx))  # True  (NOT z is true)

ctx2 = {'x': True, 'y': True, 'z': True}
print(expr.interpret(ctx2))  # True  (x AND y is true)

File: src/content/docs/topics/cs/design-patterns/interpreter.ts
type Context = Map<string, boolean>;

interface AbstractExpression {
  interpret(context: Context): boolean;
}

class Variable implements AbstractExpression {
  constructor(private name: string) {}
  interpret(context: Context): boolean {
    return context.get(this.name) ?? false;
  }
}

class AndExpression implements AbstractExpression {
  constructor(
    private left: AbstractExpression,
    private right: AbstractExpression
  ) {}
  interpret(context: Context): boolean {
    return this.left.interpret(context) && this.right.interpret(context);
  }
}

class OrExpression implements AbstractExpression {
  constructor(
    private left: AbstractExpression,
    private right: AbstractExpression
  ) {}
  interpret(context: Context): boolean {
    return this.left.interpret(context) || this.right.interpret(context);
  }
}

class NotExpression implements AbstractExpression {
  constructor(private operand: AbstractExpression) {}
  interpret(context: Context): boolean {
    return !this.operand.interpret(context);
  }
}

// (x AND y) OR (NOT z)
const x = new Variable('x');
const y = new Variable('y');
const z = new Variable('z');
const expr = new OrExpression(
  new AndExpression(x, y),
  new NotExpression(z)
);

const ctx: Context = new Map([['x', true], ['y', false], ['z', false]]);
console.log(expr.interpret(ctx)); // true  (NOT z is true)

const ctx2: Context = new Map([['x', true], ['y', true], ['z', true]]);
console.log(expr.interpret(ctx2)); // true  (x AND y is true)
